lib/Backend/Commands.hs

```diff
@@ -518,26 +519,17 @@ getEntryOrDir backend path =
       where
         go :: Path -> StateT Directory (Sem r) ()
         go rootPath = do
-          nodes <- lift $ fromMaybe [] <$> listDirectoryContents backend rootPath
+          contents <- lift $ fromMaybe (DirectoryContents [] []) <$> listDirectoryContents backend rootPath
           -- TODO: run in parallel
-          forM_ nodes \node -> do
-            -- We need to find out whether `node` is a directory name or an entry name.
-            --
-            -- Here, we rely on the fact that Vault returns directory names suffixed by `/`.
-            -- and entry names without any suffix.
-            --
-            -- If `mkPathSegment` succeeds, then `node` was not suffixed by `/`,
-            -- and thus it's an entry name.
-            -- Otherwise, it's a directory name (and `mkPath` should succeed).
-            case (mkPathSegment node, mkPath node) of
-              (Right entryName, _) -> do
-                entry <- lift $ readEntry backend (Path.appendEntryName rootPath entryName)
-                case entry of
-                  Just entry -> modify' (Dir.insertEntry entry)
-                  Nothing -> pure ()
-
-              (_, Right subdir) -> go (rootPath <> subdir)
-              _ -> lift $ throw $ InternalCommandsError (InvalidEntry node)
+          forM_ (contents ^. entryNames) \entryName -> do
+            entry <- lift $ readEntry backend (Path.appendEntryName rootPath entryName)
+            case entry of
+              Just entry -> modify' (Dir.insertEntry entry)
+              Nothing -> pure ()
```

Comment:
@dcastro If I remember correctly, when we discussed line 528 
```
Nothing -> pure ()
```
you said it is not correct to just ignore unexpected `Nothing`. 

Maybe we can throw some *internalError*, cause `contents ^. entryNames` returned an entryName that couldn't be read properly by `readEntry`, which is unexpected.

Reply:
I just copy-pasted that code). Maybe we should really throw something here

Reply:
Yeah I remember saying that. I've been thinking about it a bit more, and maybe `pure ()` is right after all :thinking: 

So, we're traversing a directory, `listDirectoryContents` gave us a list of entry names, and then `readEntry` returned 404 for one of those entry names.

That, to me, suggests that some other user deleted an entry from `vault` (or whatever backend) _while_ we were traversing the directory.

So, there's two things we can do here:
1. ignore this entry that has been concurrently deleted
2. throw and crash

(2) would imply that `coffer` can only traverse a vault directory if _no one_ else is messing with it. I don't think we want to do this, we want to allow multiple people to interact with vault at the same time. So ignoring the entry is the sensible thing to do here, IMO.

Reply:
I do think we should add a comment explaining what this branch means.

``` 
Nothing ->
  -- This entry has been concurrently deleted (e.g. by some other user) _while_ we're traversing the directory.
  -- We should just ignore it.
  pure ()
```

Reply:
Done